### DISCUSSION TITLE
Using PSFalcon with PowerShell in constrained language mode

### DISCUSSION THREAD

**Discussion by DrBob5188** [CrowdStrike/psfalcon#298] [Q&A] [answered]
When testing the PSFalcon module with a low privilege account, PowerShell was in constrained language mode, and I had difficulty importing the module.  I tracked it down to a line in the module manifest   `ScriptsToProcess     = @('Class/Class.ps1')`
It seems redundant given that the PSFalcon.psm1 file already does the same thing, although in the module context not in the user's context.
Everything works fine with that line commented out, so I'm wondering is there any potential problem I may hit in the future or if this was just added as a bit of redundancy to ensure the class loads?

I installed the module into C:\Program Files\WindowsPowerShell\Modules\PSFalcon\2.2.4\
The error message was
import-module : Importing *.ps1 files as modules is not allowed in ConstrainedLanguage mode.
At line:1 char:1
  + import-module PSFalcon
+ ~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (:) [Import-Module], InvalidOperationException
    + FullyQualifiedErrorId : Modules_ImportPSFileNotAllowedInConstrainedLanguage,Microsoft.PowerShell.Commands.ImportModuleCo

**Comment by bk-cs** [comment, marked as answer]
The Class.ps1 file loads critical functions for the operation of the module, including the loading of required libraries and the ability to make requests to the APIs.

I've never attempted to have PSFalcon work in constrained language mode. I'll see if there's anything that I can do to make it compatible, but I suspect that it won't work based on how the module is currently designed.

**Reply by DrBob5188** [reply]
Thanks for taking the time to respond.  The module is amazing!

For what it's worth, the way I got it  running in constrained language mode was to install it to a module path like C:\Program Files\WindowsPowerShell\Modules  and then comment out the ScriptsToProcess line in the manifest.
I've been using it happily from constrained language mode for a week or so without any issues.

The PSFalcon.psm1 file already loads the class file.  I just had to comment out the ScriptsToProcess     = @('Class/Class.ps1') line in the manifest to prevent it running initially in the user context.

This [reference for ScriptsToProcess](https://learn.microsoft.com/en-us/powershell/scripting/developer/module/how-to-write-a-powershell-module-manifest?view=powershell-5.1) says "Script (.ps1) files that are run in the caller's session state when the module is imported. This could be the global session state or, for nested modules, the session state of another module. ...."

I apologise for wording the initial question poorly but I was just wondering why you loaded the class file both from the PSFalcon.psm1 module file and by using the ScriptsToProcess directive?

**Reply by bk-cs** [reply]
Thanks for the clarification. I didn't realize it was being loaded twice. After some quick testing I've confirmed the same thing you saw, that it isn't required under `ScriptsToProcess` in `PSFalcon.psd1` because it's loaded as part of `PSFalcon.psm1`. I've removed it from `PSFalcon.psd1` for the next release which should ensure compatibility with constrained language mode.
